arelle/plugin/validate/DBA/rules/fr.py

```diff
@@ -177,6 +179,51 @@ def rule_fr10(
         )
 
 
+@validation(
+    hook=ValidationHook.XBRL_FINALLY,
+)
+def rule_fr21(
+        pluginData: PluginValidationDataExtension,
+        val: ValidateXbrl,
+        *args: Any,
+        **kwargs: Any,
+) -> Iterable[Validation]:
+    """
+    DBA.FR21: Assets (BalanceSheet:Assets) must equal liabilities (BalanceSheet:LiabilitiesAndEquity).
+    The check only looks at instances without dimensions and instances that only have the dimension group
+    (ConsolidatedMember) in context. Occurrences of assets and liabilities with other dimensions do not lift the
+    reporting obligation. The comparison takes place within pairs of numbers for currency and group/parent, respectively.
+    """
+    assetFacts = getFactsWithDimension(
+        val, pluginData.assetsQn, pluginData.consolidatedSoloDimensionQn,
+        pluginData.consolidatedSoloDimensionQn
+    )
+    liabilityFacts = getFactsWithDimension(
+        val, pluginData.liabilitiesQn, pluginData.consolidatedSoloDimensionQn,
+        pluginData.consolidatedSoloDimensionQn
+    )
+    foundFacts = assetFacts.union(liabilityFacts)
+    groupedFacts = defaultdict(list)
+    for fact in foundFacts:
+        groupedFacts[fact.contextID].append(fact)
+
+    for contextID, facts in groupedFacts.items():
+        assetFact = None
+        liabilityFact = None
+        for fact in facts:
+            if fact.qname == pluginData.assetsQn:
+                assetFact = fact
+            elif fact.qname == pluginData.liabilitiesQn:
+                liabilityFact = fact
+        if assetFact is not None and liabilityFact is not None:
+            if assetFact.xValue != liabilityFact.xValue:
```

Comment:
An alternative would be to collect each `fact.xValue` from `facts` into a set and check if the length is 1. This would have the benefit of handling a situation where there are potentially two liabilities facts with different values, but if we know that situation would fire an error elsewhere then we don't need to handle it here.

Reply:
Two different values for the facts would have different contextIDs or they would be duplicate facts, wouldn't they?

Reply:
I think so, but I don't think duplicate facts raise errors unless that feature is specifically invoked. See `validateDuplicateFacts`.